### PR TITLE
Fix preemption race conditions on heavily utilized nodes for e2e tests

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
> /kind bug

**What this PR does / why we need it**:
This rebases the changes from https://github.com/kubernetes/kubernetes/pull/76663, which attempts to fix up preemption e2e tests by calculating the *actual* node usage (by looking at the pods on the node) and requesting a percentage of the remaining amount instead of just what is reported by the node itself. If there are not enough actual resources available to perform the test, the test is skipped.

This also picks in fixes changes that we found were required to make this work from https://github.com/openshift/origin/pull/23645

**Special notes for your reviewer**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -24,14 +24,15 @@ import (
 	"k8s.io/client-go/tools/cache"
 
 	appsv1 "k8s.io/api/apps/v1"
-	v1 "k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
```
Comment (draveness) [inline test/e2e/scheduling/preemption.go]: Let's revert this change since import `core/v1` as `v1` is a convention in k/k, and I don't see the value of updating the alias in this PR which could bring many noises.

**Review by draveness — CHANGES_REQUESTED** [review]

**Review by draveness — COMMENTED** [review]
/kind bug

File: test/e2e/scheduling/preemption.go [hunk 2]
```diff
@@ -99,80 +100,126 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
+		pods := make([]*corev1.Pod, 0)
+		allPods, err := cs.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
+		framework.ExpectNoError(err)
 		for i, node := range nodeList.Items {
+			currentCpuUsage, currentMemUsage := getCurrentPodUsageOnTheNode(node.Name, allPods.Items, podRequestedResource)
+			framework.Logf("Current cpu and memory usage %v, %v", currentCpuUsage, currentMemUsage)
 			cpuAllocatable, found := node.Status.Allocatable["cpu"]
-			framework.ExpectEqual(found, true)
-			milliCPU := cpuAllocatable.MilliValue() * 40 / 100
+			gomega.Expect(found).To(gomega.Equal(true))
```
Comment (SataQiu) [inline test/e2e/scheduling/preemption.go]: Emm.  As #79686 mentioned, we prefer to use `framework.ExpectEqual`.

File: test/e2e/scheduling/preemption.go [hunk 3]
```diff
@@ -99,80 +100,126 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
+			milliCPU := cpuAllocatable.MilliValue()
+			milliCPU = int64(float64(milliCPU-currentCpuUsage) * float64(0.6))
 			memAllocatable, found := node.Status.Allocatable["memory"]
-			framework.ExpectEqual(found, true)
-			memory := memAllocatable.Value() * 60 / 100
-			podRes = v1.ResourceList{}
-			podRes[v1.ResourceCPU] = *resource.NewMilliQuantity(int64(milliCPU), resource.DecimalSI)
-			podRes[v1.ResourceMemory] = *resource.NewQuantity(int64(memory), resource.BinarySI)
+			gomega.Expect(found).To(gomega.Equal(true))
```
Comment (SataQiu) [inline test/e2e/scheduling/preemption.go]: ditto

File: test/e2e/scheduling/preemption.go [hunk 4]
```diff
@@ -99,80 +100,126 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
 				PriorityClassName: priorityName,
-				Resources: &v1.ResourceRequirements{
+				Resources: &corev1.ResourceRequirements{
 					Requests: podRes,
 				},
-			})
+				NodeName: node.Name,
+			}))
+			if len(pods) < 2 {
+				framework.Skipf("We need atleast two pods to be created but" +
```
Comment (SataQiu) [inline test/e2e/scheduling/preemption.go]:
```suggestion
				framework.Skipf("We need at least two pods to be created but" +
```

File: test/e2e/scheduling/preemption.go [hunk 5]
```diff
@@ -99,80 +100,126 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
+		pods := make([]*corev1.Pod, 0)
+		allPods, err := cs.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
+		gomega.Expect(err).To(gomega.BeNil())
 		for i, node := range nodeList.Items {
+			currentCpuUsage, currentMemUsage := getCurrentPodUsageOnTheNode(node.Name, allPods.Items, podRequestedResource)
+			framework.Logf("Current cpu usage and memory usage is %v, %v", currentCpuUsage, currentMemUsage)
 			cpuAllocatable, found := node.Status.Allocatable["cpu"]
-			framework.ExpectEqual(found, true)
-			milliCPU := cpuAllocatable.MilliValue() * 40 / 100
+			gomega.Expect(found).To(gomega.Equal(true))
```
Comment (SataQiu) [inline test/e2e/scheduling/preemption.go]: ditto

File: test/e2e/scheduling/preemption.go [hunk 6]
```diff
@@ -99,80 +100,126 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
+			milliCPU := cpuAllocatable.MilliValue()
+			milliCPU = int64(float64(milliCPU-currentCpuUsage) * float64(0.6))
 			memAllocatable, found := node.Status.Allocatable["memory"]
-			framework.ExpectEqual(found, true)
-			memory := memAllocatable.Value() * 60 / 100
-			podRes = v1.ResourceList{}
-			podRes[v1.ResourceCPU] = *resource.NewMilliQuantity(int64(milliCPU), resource.DecimalSI)
-			podRes[v1.ResourceMemory] = *resource.NewQuantity(int64(memory), resource.BinarySI)
+			gomega.Expect(found).To(gomega.Equal(true))
```
Comment (SataQiu) [inline test/e2e/scheduling/preemption.go]: ditto

**Review by SataQiu — COMMENTED** [review]
add some comments

File: test/e2e/scheduling/preemption.go (hunk 1 above)
Comment (damemi) [reply test/e2e/scheduling/preemption.go]: Ok, just wasn't sure why it was even aliased to `v1` in the first place

**Comment by damemi** [comment]
bump, this is green now... any more review?

**Comment by damemi** [comment]
@smarterclayton I believe these are the changes you said should be added in https://github.com/kubernetes/kubernetes/pull/80545#issuecomment-558479700. These are already carried in origin. Do you have any feedback on it? This has been dormant for a while without review

File: test/e2e/scheduling/preemption.go [hunk 7]
```diff
@@ -118,21 +130,37 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
 				Resources: &v1.ResourceRequirements{
 					Requests: podRes,
 				},
-			})
+				NodeName: node.Name,
+			}))
 			framework.Logf("Created pod: %v", pods[i].Name)
 		}
+		if len(pods) < 2 {
+			framework.Skipf("We need at least two pods to be created but" +
```
Comment (spiffxp) [inline test/e2e/scheduling/preemption.go]: I think this should be a fail instead of a skip.  If we're looking to move this test to conformance eventually, conformance tests can't skip

File: test/e2e/scheduling/preemption.go [hunk 8]
```diff
@@ -118,21 +130,37 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
 			if i == 0 {
 				priorityName = lowPriorityClassName
 			}
-			pods[i] = createPausePod(f, pausePodConfig{
+			pods = append(pods, createPausePod(f, pausePodConfig{
```
Comment (spiffxp) [inline test/e2e/scheduling/preemption.go]: if the first node is skipped due to utilization, we never use `lowPriorityClassName`, where I think the intent is first pod created should use it regardless of which node it's scheduled to? maybe this should be checking against `len(pods)` instead

**Review by spiffxp — COMMENTED** [review]
overall looks good, two nits

File: test/e2e/scheduling/preemption.go (hunk 8 above)
Comment (damemi) [reply test/e2e/scheduling/preemption.go]: I think the intent here was that the first pod should use it while all the other pods use medium priority. So in the case that the first pod doesn't get created, we do a bunch of checks to see if we're looking for a low priority a medium priority pod to be evicted.

But all that complicates the test imo. Could we just try to make 3 pods that are *all* low priority? @ravisantoshgudimetla, do you remember why we were doing it this way?

**Comment by damemi** [comment]
@spiffxp I thought some more about this and have a couple ideas. I think the problem with these e2es is that they are trying to cover every case. We tried adding this with the goal of (in the ideal case) creating a low priority pod and 2 medium priority pods, so that we know the first pod should be evicted and the other 2 remain.

However this breaks when the first pod doesn't get created, so then we see if any of the medium priority pods got evicted. I think it would be simpler to just create 3 low priority pods and check if any of those got evicted. 

There is also a problem where we create the high priority evictor-pod but don't specify a node for it to go on. So, if we have 3 nodes but only 2 lower-priority pods got created, there's a chance the high priority pod could land on the 3rd node and not evict anything.

@ahg-g do you have any thoughts on this? Could we pick a node that we know a pod is on and specify that node when we create the evictor pod [here](https://github.com/kubernetes/kubernetes/pull/82350/files#diff-e35316a06970f9000b7b154addabda8cR166-R172)? I also don't see a need for different low and medium priority pods in this test

**Comment by damemi** [comment]
@spiffxp I did make the changes you suggested to have the first pod in the list be low priority, which cleans up the code around checking medium priority pods. I do like that. I also tried manually setting the evictor pod to the node of the low priority pod, so with what I said above in mind this is ready for more review/suggestions

**Comment by spiffxp** [comment]
/retest
I'm happy with this, I'd like to see input from @smarterclayton re: whether this addresses the concerns in https://github.com/kubernetes/kubernetes/pull/80545#issuecomment-558479700

**Comment by damemi** [comment]
@spiffxp thanks, I've squashed those commits down then

**Comment by spiffxp** [comment]
/lgtm
but looking to @smarterclayton for the /approve

File: test/e2e/scheduling/preemption.go [hunk 9]
```diff
@@ -103,57 +104,77 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
...
 				Resources: &v1.ResourceRequirements{
 					Requests: podRes,
 				},
-			})
+				NodeName: node.Name,
+			}))
 			framework.Logf("Created pod: %v", pods[i].Name)
 		}
+		if len(pods) < 2 {
+			framework.Failf("We need at least two pods to be created but" +
```
Comment (smarterclayton) [inline test/e2e/scheduling/preemption.go]: I think this is fine, someone running this test on heavily utilized clusters will at least get a clear reason.  We *may* in the future want to wait a bit and retry, but we can always measure the impact of this.

File: test/e2e/scheduling/preemption.go [hunk 10]
```diff
@@ -516,3 +562,19 @@ func waitForPreemptingWithTimeout(f *framework.Framework, pod *v1.Pod, timeout t
...
+func getCurrentPodUsageOnTheNode(nodeName string, pods []v1.Pod, resource *v1.ResourceRequirements) (int64, int64) {
+	totalRequestedCPUResource := resource.Requests.Cpu().MilliValue()
+	totalRequestedMemResource := resource.Requests.Memory().Value()
+	for _, pod := range pods {
+		if pod.Spec.NodeName == nodeName {
+			if v1qos.GetPodQOS(&pod) == v1.PodQOSBestEffort {
+				continue
+			}
+		}
+		result := getNonZeroRequests(&pod)
```
Comment (smarterclayton) [inline test/e2e/scheduling/preemption.go]: This doesn't take into account init containers, so there's a small chance that this fails because a larger init container than regular container is running.  I think it's ok for now, but we should fix this.  Resources used for a pod is `max(max(init_container usage), sum(container usage))`

**Review by soltysh — APPROVED** [review]
/lgtm

File: test/e2e/scheduling/preemption.go (hunk 10 above)
Comment (ingvagabund) [reply test/e2e/scheduling/preemption.go]: Why non-zero requests? Why not just ignore resource requests which are not set?

File: test/e2e/scheduling/preemption.go [hunk 11]
```diff
@@ -516,3 +562,19 @@ func waitForPreemptingWithTimeout(f *framework.Framework, pod *v1.Pod, timeout t
 	})
 	framework.ExpectNoError(err, "pod %v/%v failed to preempt other pods", pod.Namespace, pod.Name)
 }
+
+func getCurrentPodUsageOnTheNode(nodeName string, pods []v1.Pod, resource *v1.ResourceRequirements) (int64, int64) {
+	totalRequestedCPUResource := resource.Requests.Cpu().MilliValue()
+	totalRequestedMemResource := resource.Requests.Memory().Value()
+	for _, pod := range pods {
+		if pod.Spec.NodeName == nodeName {
```
[the remark this answers is not in the document]
Comment (damemi) [reply test/e2e/scheduling/preemption.go]: Yes, good catch thank you. Pushed

File: test/e2e/scheduling/preemption.go (hunk 10 above)
Comment (damemi) [reply test/e2e/scheduling/preemption.go]: Sorry, what is the difference between them? It seems like this function is essentially just trying to get the requests, not caring about zero (or unset) because they don't add into the usage. Is that right?

Comment (ingvagabund) [reply test/e2e/scheduling/preemption.go]: `GetNonzeroRequests` which is called by `getNonZeroRequests` and implemented in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/non_zero.go adds +100m or 200MB in case cpu/memory resource requests is not set. So `getNonZeroRequests` can give you higher usage on a node that it is.
